File: Modules/LoadSingleImage.m
function handles = LoadSingleImage(handles)

% Help for the Load Single Image module:
% Category: File Processing
%
% SHORT DESCRIPTION:
% Loads a single image, which will be used for all image cycles.
% *************************************************************************
% Note: for most purposes, you will probably want to use the Load Images
% module, not this one.
%
% Tells CellProfiler where to retrieve a single image and gives the image a
% meaningful name for the other modules to access.  The module only
% functions the first time through the pipeline, and thereafter the image
% is accessible to all subsequent cycles being processed. This is
% particularly useful for loading an image like an Illumination correction
% image to be used by the CorrectIllumination_Apply module. Note: Actually,
% you can load four 'single' images using this module.
%
% Relative pathnames can be used. For example, on the Mac platform you
% could leave the folder where images are to be loaded as '.' to choose the
% default image folder, and then enter ../Imagetobeloaded.tif as the name
% of the file you would like to load in order to load the image from the
% directory one above the default image directory. Or, you could type
% .../AnotherSubfolder (note the three periods: the first is interpreted as
% a standin for the default image folder) as the folder from which images
% are to be loaded and enter the filename as Imagetobeloaded.tif to load an
% image from a different subfolder of the parent of the default image
% folder.
%
% If more than four single images must be loaded, more than one Load Single
% Image module can be run sequentially. Running more than one of these
% modules also allows images to be retrieved from different folders.
%
% See also LoadImages.

% CellProfiler is distributed under the GNU General Public License.
% See the accompanying file LICENSE for details.
%
% Developed by the Whitehead Institute for Biomedical Research.
% Copyright 2003,2004,2005.
%
% Authors:
%   Anne E. Carpenter
%   Thouis Ray Jones
%   In Han Kang
%   Ola Friman
%   Steve Lowe
%   Joo Han Chang
%   Colin Clarke
%   Mike Lamprecht
%   Susan Ma
%   Wyman Li
%
% Website: http://www.cellprofiler.org
%
% $Revision$

%%%%%%%%%%%%%%%%%
%%% VARIABLES %%%
%%%%%%%%%%%%%%%%%
drawnow

[CurrentModule, CurrentModuleNum, ModuleName] = CPwhichmodule(handles);

%textVAR01 = This module loads one image for *all* cycles that will be processed. Normally, use the Load Images module to load new sets of images during each cycle of processing.

%pathnametextVAR02 = Enter the path name to the folder where the images to be loaded are located.  Type period (.) for the default image folder.
Pathname = char(handles.Settings.VariableValues{CurrentModuleNum,2});

%filenametextVAR03 = What image file do you want to load? Include the extension, like .tif
TextToFind{1} = char(handles.Settings.VariableValues{CurrentModuleNum,3});

%textVAR04 = What do you want to call that image?
%defaultVAR04 = OrigBlue
%infotypeVAR04 = imagegroup indep
ImageName{1} = char(handles.Settings.VariableValues{CurrentModuleNum,4});

%filenametextVAR05 = What image file do you want to load? Include the extension, like .tif
TextToFind{2} = char(handles.Settings.VariableValues{CurrentModuleNum,5});

%textVAR06 = What do you want to call that image?
%defaultVAR06 = Do not load
%infotypeVAR06 = imagegroup indep
ImageName{2} = char(handles.Settings.VariableValues{CurrentModuleNum,6});

%filenametextVAR07 = What image file do you want to load? Include the extension, like .tif
TextToFind{3} = char(handles.Settings.VariableValues{CurrentModuleNum,7});

%textVAR08 = What do you want to call that image?
%defaultVAR08 = Do not load
%infotypeVAR08 = imagegroup indep
ImageName{3} = char(handles.Settings.VariableValues{CurrentModuleNum,8});

%filenametextVAR09 = What image file do you want to load? Include the extension, like .tif
TextToFind{4} = char(handles.Settings.VariableValues{CurrentModuleNum,9});

%textVAR10 = What do you want to call that image?
%defaultVAR10 = Do not load
%infotypeVAR10 = imagegroup indep
ImageName{4} = char(handles.Settings.VariableValues{CurrentModuleNum,10});

%textVAR11 = If an image slot is not being used, type a slash  /  in the box.

%textVAR11 = If your file names do not have extensions, choose the file format of the images (Note, this doesn't work currently)
%choiceVAR11 = mat
%choiceVAR11 = bmp
%choiceVAR11 = gif
%choiceVAR11 = jpg
%choiceVAR11 = tif
%choiceVAR11 = DIB
FileFormat = char(handles.Settings.VariableValues{CurrentModuleNum,11});
%inputtypeVAR11 = popupmenu

%%%VariableRevisionNumber = 3

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%% PRELIMINARY CALCULATIONS %%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
drawnow

%%% Determines which cycle is being analyzed.
SetBeingAnalyzed = handles.Current.SetBeingAnalyzed;

%%% Remove slashes '/' from the input
tmp1 = {};
tmp2 = {};
for n = 1:4
    if ~strcmp(TextToFind{n}, 'NO FILE LOADED') && ~strcmp(ImageName{n}, 'Do not load')
        tmp1{end+1} = TextToFind{n};
        tmp2{end+1} = ImageName{n};
    end
end
TextToFind = tmp1;
ImageName = tmp2;

%%% Get the pathname and check that it exists
if strncmp(Pathname,'.',1)
    if length(Pathname) == 1
        Pathname = handles.Current.DefaultImageDirectory;
    else
        Pathname = fullfile(handles.Current.DefaultImageDirectory,Pathname(2:end));
    end
end
SpecifiedPathname = Pathname;
if ~exist(SpecifiedPathname,'dir')
    error(['Image processing was canceled in the ', ModuleName, ' module because the directory "',SpecifiedPathname,'" does not exist. Be sure that no spaces or unusual characters exist in your typed entry and that the pathname of the directory begins with / (for Mac/Unix) or \ (for PC).'])
end

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%% FIRST CYCLE FILE HANDLING %%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
drawnow

if isempty(ImageName)
    error(['Image processing was canceled in the ', ModuleName, ' module because you have not chosen any images to load.'])
end

for n = 1:length(ImageName)
    %%% This try/catch will catch any problems in the load images module.
    try
        CurrentFileName = TextToFind{n};
        %%% The following runs every time through this module (i.e. for
        %%% every cycle).
        %%% Saves the original image file name to the handles
        %%% structure.  The field is named appropriately based on
        %%% the user's input, in the Pipeline substructure so that
        %%% this field will be deleted at the end of the analysis
        %%% batch.
        fieldname = ['Filename', ImageName{n}];
        handles.Pipeline.(fieldname) = CurrentFileName;
        fieldname = ['Pathname', ImageName{n}];
        handles.Pipeline.(fieldname) =  Pathname;

        FileAndPathname = fullfile(Pathname, CurrentFileName);
        if strcmpi(FileFormat,'mat')
            try StructureLoadedImage = load(FileAndPathname);
                LoadedImage = StructureLoadedImage.Image;
            catch
                try [LoadedImage, handles] = CPimread(FileAndPathname,handles);
                catch
                    error(['Image processing was canceled in the ', ModuleName, ' module because an error occurred when trying to load this image: ',FileAndPathname,'. The file may be corrupt.']);
                end
            end
        else [LoadedImage, handles] = CPimread(FileAndPathname,handles);
        end
        %%% Saves the image to the handles structure.
        handles.Pipeline.(ImageName{n}) = LoadedImage;

    catch ErrorMessage = lasterr;
        ErrorNumber = {'first','second','third','fourth'};
        error(['Image processing was canceled in the ', ModuleName, ' module because an error occurred when trying to load the ', ErrorNumber{n}, ' set of images. Please check the settings. A common problem is that there are non-image files in the directory you are trying to analyze, or that the image file is not in the format you specified: ', FileFormat, '. Matlab says the problem is: ', ErrorMessage])
    end % Goes with: catch

    % Create a cell array with the filenames
    FileNames(n) = {CurrentFileName};
end

%%% -- Save to the handles.Measurements structure for reference in output files --------------- %%%
%%% NOTE: The structure for filenames and pathnames will be a cell array of cell arrays

%%% First, fix feature names and the pathname
PathNames = cell(1,length(ImageName));
FileNamesText = cell(1,length(ImageName));
PathNamesText = cell(1,length(ImageName));
for n = 1:length(ImageName)
    PathNames{n} = Pathname;
    FileNamesText{n} = ['Filename ', ImageName{n}];
    PathNamesText{n} = ['Path ', ImageName{n}];
end

%%% Since there may be several load modules in the pipeline which all write to the
%%% handles.Measurements.Image.FileName field, we have store filenames in an "appending" style.
%%% Here we check if any of the modules above the current module in the pipline has written to
%%% handles.Measurements.Image.Filenames. Then we should append the current filenames and path
%%% names to the already written ones.
if  isfield(handles,'Measurements') && isfield(handles.Measurements,'Image') && isfield(handles.Measurements.Image,'FileNames')
    if length(handles.Measurements.Image.FileNames) == SetBeingAnalyzed
        % Get existing file/path names. Returns a cell array of names
        ExistingFileNamesText = handles.Measurements.Image.FileNamesText;
        ExistingFileNames     = handles.Measurements.Image.FileNames{SetBeingAnalyzed};
        ExistingPathNamesText = handles.Measurements.Image.PathNamesText;
        ExistingPathNames     = handles.Measurements.Image.PathNames{SetBeingAnalyzed};

        % Append current file names to existing file names
        FileNamesText = cat(2,ExistingFileNamesText,FileNamesText);
        FileNames     = cat(2,ExistingFileNames,FileNames);
        PathNamesText = cat(2,ExistingPathNamesText,PathNamesText);
        PathNames     = cat(2,ExistingPathNames,PathNames);
    end
end

%%% Write to the handles.Measurements.Image structure
handles.Measurements.Image.FileNamesText                   = FileNamesText;
handles.Measurements.Image.FileNames(SetBeingAnalyzed)     = {FileNames};
handles.Measurements.Image.PathNamesText                   = PathNamesText;
handles.Measurements.Image.PathNames(SetBeingAnalyzed)     = {PathNames};

%%%%%%%%%%%%%%%%%%%%%%%
%%% DISPLAY RESULTS %%%
%%%%%%%%%%%%%%%%%%%%%%%
drawnow

%%% The figure window display is unnecessary for this module, so the figure
%%% window is closed the first time through the module.
%%% Determines the figure number.
ThisModuleFigureNumber = handles.Current.(['FigureNumberForModule',CurrentModule]);
%%% Closes the window if it is open.
if any(findobj == ThisModuleFigureNumber)
    if handles.Current.SetBeingAnalyzed == handles.Current.StartingImageSet
        CPresizefigure('','NarrowText')
    end
    for n = 1:length(ImageName)
        drawnow
        %%% Activates the appropriate figure window.
        CPfigure(handles,'Text',ThisModuleFigureNumber);
        if iscell(ImageName)
            TextString = [ImageName{n},': ',FileNames{n}];
        else
            TextString = [ImageName,': ',FileNames];
        end
        uicontrol('style','text','units','normalized','fontsize',handles.Preferences.FontSize,'HorizontalAlignment','left','string',TextString,'position',[.05 .85-(n-1)*.15 .95 .1],'BackgroundColor',[.7 .7 .9])
    end
end

File: pyCellProfiler/cellprofiler/modules/loadsingleimage.py
"""<b>LoadSingleImage</b> loads a single image, which will be used for all image cycles.
<hr>
Note: for most purposes, you will probably want to use the Load Images
module, not this one.

Tells CellProfiler where to retrieve a single image and gives the image a
meaningful name for the other modules to access.  The module only
executes the first time through the pipeline, and thereafter the image
is accessible to all subsequent cycles being processed. This is
particularly useful for loading an image like an Illumination correction
image to be used by the CorrectIllumination_Apply module. Note: Actually,
you can load four 'single' images using this module.

See also <b>LoadImages</b>.

"""
__version__="Revision: $1 "

# CellProfiler is distributed under the GNU General Public License.
# See the accompanying file LICENSE for details.
# 
# Developed by the Broad Institute
# Copyright 2003-2010
# 
# Please see the AUTHORS file for credits.
# 
# Website: http://www.cellprofiler.org

import hashlib
import numpy as np
import re
import os

import cellprofiler.cpimage as cpi
import cellprofiler.cpmodule as cpm
import cellprofiler.measurements as cpmeas
import cellprofiler.preferences as cpprefs
import cellprofiler.settings as cps
from loadimages import LoadImagesImageProvider

DIR_DEFAULT_IMAGE_FOLDER = "Default input folder"
DIR_DEFAULT_OUTPUT_FOLDER = "Default output folder"
DIR_CUSTOM_FOLDER = "Custom folder"
DIR_CUSTOM_WITH_METADATA = "Custom with metadata"

class LoadSingleImage(cpm.CPModule):

    module_name = "LoadSingleImage"
    category = "File Processing"
    variable_revision_number = 1
    def create_settings(self):
        """Create the settings during initialization
        
        """
        self.dir_choice = cps.Choice(
            "Which folder contains the image files?",
            [DIR_DEFAULT_IMAGE_FOLDER,
             DIR_DEFAULT_OUTPUT_FOLDER,
             DIR_CUSTOM_FOLDER,
             DIR_CUSTOM_WITH_METADATA], 
            doc = '''It is best to store your illumination function
            in either the input or output folder, so that the correct image is loaded into 
            the pipeline and typos are avoided.  If you must store it in another folder, 
            select 'Custom'. You can use metadata from the image set to
            construct the name by selecting 'Custom with metadata'.''')
        self.custom_directory = cps.Text(
            "What is the name of the folder containing the image files?",".",
            doc='''This is the name of the folder that has your illumination
            files in it. If you chose "Custom with metadata" above, you can
            specify a path that is based on metadata associated with the
            image set. For instance, if you have a "Plate" metadata element,
            you can specify a path name of "./\g&lt;Plate&gt;" to get files
            from the folder associated with your image's plate.
            You can prefix the folder name with "." (a period) to make the 
            root folder the default input folder or "&" (an ampersand) 
            to make the root folder the default output folder.''')
        self.file_settings = []
        self.add_file(can_remove = False)
        self.add_button = cps.DoSomething("", "Add another image", self.add_file)

    def add_file(self, can_remove = True):
        """Add settings for another file to the list"""
        group = cps.SettingsGroup()
        if can_remove:
            group.append("divider", cps.Divider(line=False))
        def get_directory_fn():
            if self.dir_choice == DIR_DEFAULT_IMAGE_FOLDER:
                return cpprefs.get_default_image_directory()
            elif self.dir_choice == DIR_DEFAULT_OUTPUT_FOLDER:
                return cpprefs.get_default_output_directory()
            elif self.dir_choice == DIR_CUSTOM_FOLDER:
                return self.custom_directory.value
            return os.curdir()
        
        group.append("file_name", cps.FilenameText(
            "Filename of the image to load (Include the extension, e.g., .tif)",
            "None",
            get_directory_fn = get_directory_fn,
            exts = [("Tagged image file (*.tif)","*.tif"),
                    ("Portable network graphics (*.png)", "*.png"),
                    ("JPEG file (*.jpg)", "*.jpg"),
                    ("Bitmap file (*.bmp)", "*.bmp"),
                    ("GIF file (*.gif)", "*.gif"),
                    ("Matlab image (*.mat)","*.mat"),
                    ("All files (*.*)", "*.*")] ))
        group.append("image_name", cps.FileImageNameProvider("What do you want to call that image?", "OrigBlue"))
        if can_remove:
            group.append("remove", cps.RemoveSettingButton("", "Remove this image", self.file_settings, group))
        self.file_settings.append(group)

    def settings(self):
        """Return the settings in the order in which they appear in a pipeline file"""
        result = [self.dir_choice, self.custom_directory]
        for file_setting in self.file_settings:
            result += [file_setting.file_name, file_setting.image_name]
        return result

    def visible_settings(self):
        result = [self.dir_choice]
        if self.dir_choice in (DIR_CUSTOM_FOLDER, DIR_CUSTOM_WITH_METADATA):
            result += [self.custom_directory]
        for file_setting in self.file_settings:
            result += file_setting.visible_settings()
        result.append(self.add_button)
        return result 

    def prepare_settings(self, setting_values):
        """Adjust the file_settings depending on how many files there are"""
        count = (len(setting_values)-2)/2
        del self.file_settings[count:]
        while len(self.file_settings) < count:
            self.add_file()

    def prepare_to_create_batch(self, pipeline, image_set_list, fn_alter_path):
        '''Prepare to create a batch file
        
        This function is called when CellProfiler is about to create a
        file for batch processing. It will pickle the image set list's
        "legacy_fields" dictionary. This callback lets a module prepare for
        saving.
        
        pipeline - the pipeline to be saved
        image_set_list - the image set list to be saved
        fn_alter_path - this is a function that takes a pathname on the local
                        host and returns a pathname on the remote host. It
                        handles issues such as replacing backslashes and
                        mapping mountpoints. It should be called for every
                        pathname stored in the settings or legacy fields.
        '''
        if self.dir_choice == DIR_DEFAULT_IMAGE_FOLDER:
            self.dir_choice.value = DIR_CUSTOM_FOLDER
            self.custom_directory.value = cpprefs.get_default_image_directory()
        elif self.dir_choice == DIR_DEFAULT_OUTPUT_FOLDER:
            self.dir_choice.value = DIR_CUSTOM_FOLDER
            self.custom_directory.value = cpprefs.get_default_output_directory()
        elif self.dir_choice == DIR_CUSTOM_FOLDER:
            self.custom_directory.value = cpprefs.get_absolute_path(
                self.custom_directory.value)
        elif self.dir_choice == DIR_CUSTOM_WITH_METADATA:
            path = self.custom_directory.value
            end_new_style = path.find("\\g<")
            end_old_style = path.find("\(?")
            end = (end_new_style 
                   if (end_new_style != -1 and 
                       (end_old_style == -1 or end_old_style > end_new_style))
                   else end_old_style)
            if end != -1:
                pre_path = path[:end]
                pre_path = cpprefs.get_absolute_path(pre_path)
                pre_path = fn_alter_path(pre_path)
                path = pre_path + path[end:]
                self.custom_directory.value = path
                return True
        self.custom_directory.value = fn_alter_path(self.custom_directory.value)
        return True

    def get_base_directory(self, workspace):
        if self.dir_choice == DIR_DEFAULT_IMAGE_FOLDER:
            base_directory = cpprefs.get_default_image_directory()
        elif self.dir_choice == DIR_DEFAULT_OUTPUT_FOLDER:
            base_directory = cpprefs.get_default_output_directory()
        elif self.dir_choice in (DIR_CUSTOM_FOLDER, DIR_CUSTOM_WITH_METADATA):
            base_directory = self.custom_directory.value
            if self.dir_choice == DIR_CUSTOM_WITH_METADATA:
                base_directory = workspace.measurements.apply_metadata(base_directory)
            if (base_directory[:2] == '.'+ os.sep or
                (os.altsep and base_directory[:2] == '.'+os.altsep)):
                # './filename' -> default_image_folder/filename
                base_directory = os.path.join(cpprefs.get_default_image_directory(),
                                              base_directory[2:])
            elif (base_directory[:2] == '&'+ os.sep or
                  (os.altsep and base_directory[:2] == '&'+os.altsep)):
                base_directory = os.path.join(cpprefs.get_default_output_directory(),
                                              base_directory[2:])
        return base_directory
    
    def get_file_names(self, workspace):
        """Get the files for the current image set
        
        workspace - workspace for current image set
        
        returns a dictionary of image_name keys and file path values
        """
        result = {}
        for file_setting in self.file_settings:
            file_pattern = file_setting.file_name.value
            file_name = workspace.measurements.apply_metadata(file_pattern)
            result[file_setting.image_name.value] = file_name
                
        return result
            
    def run(self, workspace):
        dict = self.get_file_names(workspace)
        root = self.get_base_directory(workspace)
        statistics = [("Image name","File")]
        m = workspace.measurements
        for image_name in dict.keys():
            provider = LoadImagesImageProvider(image_name, root, 
                                               dict[image_name])
            workspace.image_set.providers.append(provider)
            #
            # Add measurements
            #
            m.add_measurement('Image','FileName_'+image_name, dict[image_name])
            m.add_measurement('Image','PathName_'+image_name, root)
            pixel_data = provider.provide_image(workspace.image_set).pixel_data
            digest = hashlib.md5()
            digest.update(np.ascontiguousarray(pixel_data).data)
            m.add_measurement('Image','MD5Digest_'+image_name, digest.hexdigest())
            statistics += [(image_name, dict[image_name])]
        if workspace.frame:
            title = "Load single image: image set # %d"%(workspace.measurements.image_set_number+1)
            figure = workspace.create_or_find_figure(title=title,
                                                     subplots=(1,1))
            figure.subplot_table(0,0, statistics)
    
    def get_measurement_columns(self, pipeline):
        columns = []
        for file_setting in self.file_settings:
            image_name = file_setting.image_name.value
            columns += [(cpmeas.IMAGE, '_'.join((feature, image_name)), coltype)
                        for feature, coltype in (
                            ('FileName', cpmeas.COLTYPE_VARCHAR_FILE_NAME),
                            ('PathName', cpmeas.COLTYPE_VARCHAR_PATH_NAME),
                            ('MD5Digest', cpmeas.COLTYPE_VARCHAR_FORMAT % 32))]
        return columns
    
    def upgrade_settings(self, setting_values, variable_revision_number, module_name, from_matlab):
        if from_matlab and variable_revision_number == 4:
            new_setting_values = list(setting_values)
            # The first setting was blank in Matlab. Now it contains
            # the directory choice
            if setting_values[1] == '.':
                new_setting_values[0] = DIR_DEFAULT_IMAGE_FOLDER
            elif setting_values[1] == '&':
                new_setting_values[0] = DIR_DEFAULT_OUTPUT_FOLDER
            else:
                new_setting_values[0] = DIR_CUSTOM_FOLDER
            #
            # Remove "Do not use" images
            #
            for i in [8, 6, 4]:
                if new_setting_values[i+1] == cps.DO_NOT_USE:
                    del new_setting_values[i:i+2]
            setting_values = new_setting_values
            from_matlab = False
            variable_revision_number = 1
        #
        # Minor revision: default image folder -> default input folder
        #
        if variable_revision_number == 1 and not from_matlab:
            if setting_values[0].startswith("Default image"):
                setting_values = [DIR_DEFAULT_IMAGE_FOLDER] + setting_values[1:] 
        return setting_values, variable_revision_number, from_matlab

